h2/src/main/org/h2/command/ddl/PrepareProcedure.java

```diff
@@ -52,7 +51,7 @@ public void setPrepared(Prepared prep) {
 
     @Override
     public ArrayList<Parameter> getParameters() {
-        return Utils.newSmallArrayList();
+        return new ArrayList<>(0);
```

Comment:
It's actually little faster (1 if less) to call the parameterless constructor for ArrayList then to call new ArrayList<>(0).

It's probably even better to use Collections.emptyList(), provided returned lists won't be modified. For this method it doesn't seem they are. Also method would need to be return List<Parameter> instead.

Reply:
It depends on Java version.
* On Java 7 `new ArrayList<>()` initializes underlying array of size 10, `new ArrayList<>(0)` initializes underlying array of size 0.
* On Java 8 `new ArrayList<>()` does not allocate additional objects, `new ArrayList<>(0)` initializes underlying array of size 0.
* On Java 9+ both methods do not allocate additional objects.

Usage of `List` and actual returning of collections with different types will force C2 to compile different code paths for different collections, usage of `ArrayList.iterator()` is optimized well in hot methods. The overall impact may be negative, this should be tested somehow.

Reply:
Personally I don't think that performance on Java 7 is significant now. Java 7 is completely outdated.